### ISSUE TITLE
Getting "an eigenvalue has become negative" when running CMAES

### ISSUE THREAD

**Issue by mikaelgrev** [MOEAFramework/MOEAFramework#187]
It happens after a while after running CMAES optimisation and then stalls the thread and nothing more is evaluated.

I use an Executor and it is happening for a problem that shouldn't be special in any way. Doesn't matter if I have one or two objectives or if I use many or one thread. All values for CMAES are default.

Does anyone know why it happens and how to fix it? I can see the error in the CMAES.java file but setting it to 0 (as the code does) does seem to be a bad way to handle the problem since it hangs or stalls any further evaluation.

I have not found the problem to happen with any of the other algorithms.

I used the latest version of the MOEAFramework.

**Comment by dhadka** [comment]
Hi, thanks for reporting this issue.  What problem are you using?  If it's one of the built-in problems, could you please provide the code you're running so I can try and recreate the error?

It does seem unusual that this is stalling the thread.  If I can reproduce it locally, I can try to investigate why this is happening.

**Comment by mikaelgrev** [comment]
Hi! Unfortunately I can't share the problem as it would drag in a lot of code that is not shareable.

I do run the algorithm "forever" by setting a very large number of evals. Can it be that it finds a local very flat minimum and hovers around that until the math breaks down?

I'll see if I can isolate the problem or get more info about it in any way.

Thanks for trying. :)

**Comment by dhadka** [comment]
I was able to recreate the issue.

It's not caused by the negative eigenvalue or setting it to 0, although that might be a symptom of the problem.  One of the eigenvalues becomes extremely large (e.g., 2.37e10).  CMAES then uses these eigenvalues when generating the offspring, but due to the large value it's producing variables outside the min/max bounds.  This then falls into an infinite loop.

Investigating the root cause now to try and figure out what causes these negative/large eigenvalues.

**Comment by mikaelgrev** [comment]
Excellent, good catch.

I have used Commons Math's version of CMAES before, and it seems very stable. Maybe something can be lifted from that given they both originate from the same code? I switched to MOEA to get the multi-threading and an easy way to test more algorithms.

**Comment by mikaelgrev** [comment]
When I reviewed the code I saw two opportunities for improvement in CMAES.java.

1) in samplePopulation(), all variables are thrown away if one is outside bounds, and the randomisation restarts. If you have many variables and a high sigma relative to bounds, this will take a long time (possibly almost infinite). Couldn't it just re-randomize the single variable that is outside bounds, which would make it a lot faster and predictable?

2) It would be good if the sigma value, which now is just one for all variables, could instead be an array of sigmas, one for each variable. The change should be straight forward as far as I can see. The only thing is that the constructor can't be changed, and I guess you don't want another one, so it would have to be a setter instead. It is quite unlikely that sigmas for all variables would be the same, so I'm not sure why it's done that way. The Commons Math CMAES have one for each variable.

Let me know if you want a ticket for one of each of these.
